lib/src/install.rs

```diff
@@ -1596,21 +1596,31 @@ fn remove_all_in_dir_no_xdev(d: &Dir, mount_err: bool) -> Result<()> {
 }
 
 #[context("Removing boot directory content")]
-fn clean_boot_directories(rootfs: &Dir) -> Result<()> {
+fn clean_boot_directories(rootfs: &Dir, is_ostree: bool) -> Result<()> {
     let bootdir =
         crate::utils::open_dir_remount_rw(rootfs, BOOT.into()).context("Opening /boot")?;
-    // This should not remove /boot/efi note.
-    remove_all_in_dir_no_xdev(&bootdir, false)?;
-    // TODO: Discover the ESP the same way bootupd does it; we should also
-    // support not wiping the ESP.
-    if ARCH_USES_EFI {
-        if let Some(efidir) = bootdir
-            .open_dir_optional(crate::bootloader::EFI_DIR)
-            .context("Opening /boot/efi")?
-        {
-            remove_all_in_dir_no_xdev(&efidir, false)?;
+
+    if is_ostree {
+        // On ostree systems, the boot directory already has our desired format, we should only
+        // remove the bootupd-state.json file to avoid bootupctl complaining it already exists.
```

Comment:
Hmm though I think this is going to work when staying within a single OS major usually, in the *general* case I think we can get some skew here when we don't clear the ESP (as is done below for the non-ostree/bootupd case).

Related to this on the flip side, ostree and bootupd are separate, orthogonal projects and we were actually talking about using bootupd for package-based Fedora too. In that scenario, we would still need to blow away the bootupd state.

But then again *really* the right fix in the medium term here is again to basically treat this as an upgrade, which bootupd already supports. So, eh.